rllib/algorithms/appo/utils.py

```diff
@@ -22,6 +23,143 @@
+        """
+        Add a new, unseen item to the END of the buffer.
+        If full, the oldest item (from the FRONT) is evicted.
+        """
+        wrapped_item = _ImpactItem(item)
+        evicted_data = None
+
+        with self._lock:
+            self._total_puts += 1
+
+            # 1. Eviction: Evict from the FRONT if full
+            if len(self._buffer) >= self._capacity:
+                evicted_item = self._buffer.popleft()
+                evicted_data = evicted_item.data
+                self._total_evictions += 1
+                # The semaphore count stays the same because one
+                # item is removed, but another is about to be added.
+            else:
+                # We are adding a new item, so release the semaphore.
+                self._items_available.release()
+
+            # 2. Add new item to the END
+            self._buffer.append(wrapped_item)
+            return evicted_data
+
+    def get(self, block: bool = True, timeout: Optional[float] = None) -> Any:
+        """
+        Get the oldest item (from the FRONT).
+        If the item is not "spent" (reuse_count < max_reuse),
+        it is re-queued at the END of the buffer.
+        """
+        # 1. Acquire right to get an item. This blocks if empty.
+        acquired = self._items_available.acquire(blocking=block, timeout=timeout)
+        if not acquired:
+            if block:
+                raise TimeoutError("Timeout waiting for item")
+            else:
+                raise IndexError("Buffer is empty")
+
+        # 2. Get item from the FRONT (thread-safe)
+        with self._lock:
+            if not self._buffer:
+                # Safety check, should not happen if semaphore is correct
+                self._items_available.release()  # Release the sem we acquired
+                raise IndexError("Buffer is empty")
+
+            wrapped_item = self._buffer.popleft()
+            self._total_gets += 1
+
+        # 3. Process item logic *outside* the lock
+        wrapped_item.reuse_count += 1
+
+        # 4. Check for re-queue
+        if wrapped_item.reuse_count < self._max_reuse:
+            # Item is not spent, put it back at the END
+            with self._lock:
+                self._buffer.append(wrapped_item)
+
+            # Release semaphore again, as item is still available (just at the back)
+            self._items_available.release()
+        else:
+            # Item is spent. Do not re-queue. Do not release semaphore.
+            # The total number of available items has decreased by 1.
+            self._total_spent += 1
+
+        return wrapped_item.data
+
+    def qsize(self) -> int:
+        with self._lock:
+            return len(self._buffer)
+
+    def __len__(self) -> int:
+        return self.qsize()
+
+    def task_done(self):
+        pass
+
+    def get_stats(self) -> dict:
+        """Get buffer statistics for monitoring."""
+        with self._lock:
+            avg_reuse = 0
```

Comment:
Minor performance improvement through only computing the buffer length once and not creating the counts list through a generator
```
with self._lock:
    buffer_len = len(self._buffer)
    avg_reuse = -1  # -1 if no buffer as 0 is a valid possible number of reuse

    if self._buffer:
        avg_reuse = sum(item.reuse_count for item in self._buffer) / buffer_len

    return {
        "size": buffer_len,
        "capacity": self._capacity,
        "max_reuse": self._max_reuse,
        "total_puts": self._total_puts,
        "total_gets": self._total_gets,
        "total_evictions": self._total_evictions,
        "total_spent_items": self._total_spent,
        "avg_reuse_in_buffer": avg_reuse,
    }
```